File: examples/android-demo/example/src/main/java/com/tencent/mtt/hippy/example/module/turbo/TurboMap.java
package com.tencent.mtt.hippy.example.module.turbo;

import com.tencent.mtt.hippy.annotation.HippyTurboObj;
import com.tencent.mtt.hippy.annotation.HippyTurboProp;
import com.tencent.mtt.hippy.common.HippyArray;
import com.tencent.mtt.hippy.common.HippyMap;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * 支持以引用的方式暴露给js使用的Map
 * TurboMap支持的类型如下：
 * 【框架转换后透传】int/long/double/String/boolean/HippyMap/HippyArray
 * 【引用】TurboArray/TurboMap
 */
@HippyTurboObj
public class TurboMap {

  private final HashMap<String, Object> mDatas;

  @Override
  public String toString() {
    return mDatas.toString();
  }

  public TurboMap() {
    mDatas = new HashMap<>();
  }

  public boolean containsKey(String key) {
    return mDatas.containsKey(key);
  }

  @HippyTurboProp(expose = true)
  public int size() {
    return mDatas.size();
  }

  public Set<String> keySet() {
    return mDatas.keySet();
  }

  public Set<Map.Entry<String, Object>> entrySet() {
    return mDatas.entrySet();
  }

  @HippyTurboProp(expose = true)
  public String getString(String key) {
    Object value = mDatas.get(key);
    return value == null ? null : String.valueOf(value);
  }

  public void remove(String key) {
    mDatas.remove(key);
  }

  @HippyTurboProp(expose = true)
  public double getDouble(String key) {
    Object value = mDatas.get(key);
    return value instanceof Number ? ((Number) value).doubleValue() : 0;
  }

  @HippyTurboProp(expose = true)
  public int getInt(String key) {
    Object value = mDatas.get(key);
    return value instanceof Number ? ((Number) value).intValue() : 0;
  }

  @HippyTurboProp(expose = true)
  public boolean getBoolean(String key) {
    Object value = mDatas.get(key);
    return value != null && (boolean) value;
  }

  @HippyTurboProp(expose = true)
  public long getLong(String key) {
    Object value = mDatas.get(key);
    return value instanceof Number ? ((Number) value).longValue() : 0;
  }

  @HippyTurboProp(expose = true)
  public TurboMap getTurboMap(String key) {
    Object value = mDatas.get(key);
    if (value instanceof TurboMap) {
      return (TurboMap) value;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public TurboArray getTurboArray(String key) {
    Object value = mDatas.get(key);
    if (value instanceof TurboArray) {
      return (TurboArray) value;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public HippyMap getHippyMap(String key) {
    Object value = mDatas.get(key);
    if (value instanceof HippyMap) {
      return (HippyMap) value;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public HippyArray getHippyArray(String key) {
    Object value = mDatas.get(key);
    if (value instanceof HippyArray) {
      return (HippyArray) value;
    }
    return null;
  }

  public boolean isNull(String key) {
    return mDatas.get(key) == null;
  }

  public void pushNull(String key) {
    mDatas.put(key, null);
  }

  public void pushInt(String key, int value) {
    mDatas.put(key, value);
  }

  public void pushString(String key, String value) {
    mDatas.put(key, value);
  }

  public void pushBoolean(String key, boolean value) {
    mDatas.put(key, value);
  }

  public void pushDouble(String key, double value) {
    mDatas.put(key, value);
  }

  public void pushLong(String key, long value) {
    mDatas.put(key, value);
  }

  public void pushTurboArray(String key, TurboArray array) {
    mDatas.put(key, array);
  }

  public void pushTurboMap(String key, TurboMap map) {
    mDatas.put(key, map);
  }

  public void pushHippyArray(String key, HippyArray array) {
    mDatas.put(key, array);
  }

  public void pushHippyMap(String key, HippyMap map) {
    mDatas.put(key, map);
  }

  public void pushAll(Map map) {
    if (null != map) {
      mDatas.putAll(map);
    }
  }

  public void clear() {
    mDatas.clear();
  }
}

File: framework/examples/android-demo/src/main/java/com/openhippy/example/turbo/TurboMap.kt
/*
 * Tencent is pleased to support the open source community by making Hippy
 * available.
 * Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.openhippy.example.turbo

import com.tencent.mtt.hippy.annotation.HippyTurboObj
import com.tencent.mtt.hippy.annotation.HippyTurboProp
import com.tencent.mtt.hippy.common.HippyArray
import com.tencent.mtt.hippy.common.HippyMap

/**
 * A `Map` type that can be used as reference by Js
 *
 * The following data types are supported:
 * - **As primitive**: int/long/double/String/boolean/HippyMap/HippyArray
 * - **As reference**: TurboArray/TurboMap
 */
@HippyTurboObj
class TurboMap {
    private val mData: HashMap<String, Any?> = HashMap()

    override fun toString(): String = mData.toString()

    fun containsKey(key: String): Boolean = mData.containsKey(key)

    @HippyTurboProp(expose = true)
    fun size(): Int = mData.size

    fun keySet(): Set<String> = mData.keys

    fun entrySet(): Set<Map.Entry<String, Any?>> = mData.entries

    @HippyTurboProp(expose = true)
    fun getString(key: String): String? = mData[key]?.toString()

    fun remove(key: String) {
        mData.remove(key)
    }

    @HippyTurboProp(expose = true)
    fun getDouble(key: String): Double = (mData[key] as? Number)?.toDouble() ?: 0.0

    @HippyTurboProp(expose = true)
    fun getInt(key: String): Int = (mData[key] as? Number)?.toInt() ?: 0

    @HippyTurboProp(expose = true)
    fun getBoolean(key: String): Boolean = mData[key] as? Boolean ?: false

    @HippyTurboProp(expose = true)
    fun getLong(key: String): Long = (mData[key] as? Number)?.toLong() ?: 0

    @HippyTurboProp(expose = true)
    fun getTurboMap(key: String): TurboMap? = mData[key] as? TurboMap

    @HippyTurboProp(expose = true)
    fun getTurboArray(key: String): TurboArray? = mData[key] as? TurboArray

    @HippyTurboProp(expose = true)
    fun getHippyMap(key: String): HippyMap? = mData[key] as? HippyMap

    @HippyTurboProp(expose = true)
    fun getHippyArray(key: String): HippyArray? = mData[key] as? HippyArray

    fun isNull(key: String): Boolean = mData[key] == null

    fun pushNull(key: String) {
        mData[key] = null
    }

    fun pushInt(key: String, value: Int) {
        mData[key] = value
    }

    fun pushString(key: String, value: String?) {
        mData[key] = value
    }

    fun pushBoolean(key: String, value: Boolean) {
        mData[key] = value
    }

    fun pushDouble(key: String, value: Double) {
        mData[key] = value
    }

    fun pushLong(key: String, value: Long) {
        mData[key] = value
    }

    fun pushTurboArray(key: String, array: TurboArray?) {
        mData[key] = array
    }

    fun pushTurboMap(key: String, map: TurboMap?) {
        mData[key] = map
    }

    fun pushHippyArray(key: String, array: HippyArray?) {
        mData[key] = array
    }

    fun pushHippyMap(key: String, map: HippyMap?) {
        mData[key] = map
    }

    fun pushAll(map: Map<*, *>?) {
        map?.forEach { (key, value) -> mData[key.toString()] = value }
    }

    fun clear() {
        mData.clear()
    }
}